README.md
# thrreg

## Overview
thrreg estimates the threshold regression model by Hansen (2000). 
It provides three functions:
* `thr_test_hom()` tests for a threshold in a linear regression under homoskedasticity.
* `thr_test_het()` tests for a threshold in a linear regression under heteroskedasticity.
* `thr_est()` estimates the threshold and the regression parameters of the threshold model.

## Installation
Install release version from CRAN:
```r
install.packages("thrreg")
```
Install development version from GitHub:
```r
# install.packages("devtools")
devtools::install_github("mlkremer/thrreg")
```

## Usage
This example performs the empirical work reported in Hansen (2000) and prints 
the LaTeX output to `output/output.tex`:
```r
library(thrreg)

## Load data
data <- dur_john

## Define variable names for output (optional)
var_names <- c("GDP\\_Gwth", "Log\\_GDP\\_1960", "Log\\_Inv/GDP",
               "Log\\_Pop\\_Gwth", "Log\\_School", "GDP\\_1960", "Literacy")
xi <- 2:5
h <- 1


## First Level
sink("output/output.tex")
cat("\\section{First Sample Split}", "\n\n")

## Test for a Sample Split
test_1_gdp <- thr_test_het(data, "GDPGwth", xi, "GDP1960", var_names)
test_1_lit <- thr_test_het(data, "GDPGwth", xi, "Literacy", var_names)

## Estimate Sample Split
qhat_1 <- thr_est(data, "GDPGwth", xi, "GDP1960", h, test_1_gdp$p_value,
                  var_names, digits.thr = 0, header = var_names[1])
sink()


## Second Level
dat_2l <- subset(data, GDP1960 <= qhat_1)
dat_2u <- subset(data, GDP1960 > qhat_1)

sink("output/output.tex", append = T)
cat(paste0("\\section{Second Sample Split: Subsample, Incomes below ", qhat_1,
           "}", "\n\n"))

## Test for a Sample Split
test_2l_gdp <- thr_test_het(dat_2l, "GDPGwth", xi, "GDP1960", var_names)
test_2l_lit <- thr_test_het(dat_2l, "GDPGwth", xi, "Literacy", var_names)

cat(paste0("\\section{Second Sample Split: Subsample, Incomes above ", qhat_1,
           "}", "\n\n"))

## Test for a Sample Split
test_2u_gdp <- thr_test_het(dat_2u, "GDPGwth", xi, "GDP1960", var_names)
test_2u_lit <- thr_test_het(dat_2u, "GDPGwth", xi, "Literacy", var_names)

## Estimate Sample Split
qhat_2 <- thr_est(dat_2u, "GDPGwth", xi, "Literacy", h, test_2u_lit$p_value,
                  var_names, digits.thr = 0, header = var_names[1])
sink()


## Third Level
dat_3l <- subset(dat_2u, Literacy <= qhat_2)
dat_3u <- subset(dat_2u, Literacy > qhat_2)

sink("output/output.tex", append = T)
cat(paste0("\\section{Third Sample Split: Subsample, Incomes above ", qhat_1,
           ", Literacy below ", qhat_2, "}", "\n\n"))

## Test for a Sample Split
test_3l_gdp <- thr_test_het(dat_3l, "GDPGwth", xi, "GDP1960")
test_3l_lit <- thr_test_het(dat_3l, "GDPGwth", xi, "Literacy")

cat(paste0("\\section{Third Sample Split: Subsample, Incomes above ", qhat_1,
           ", Literacy above ", qhat_2, "}", "\n\n"))

## Test for a Sample Split
test_3u_gdp <- thr_test_het(dat_3u, "GDPGwth", xi, "GDP1960")
test_3u_lit <- thr_test_het(dat_3u, "GDPGwth", xi, "Literacy")

sink()
```

Compile LaTeX output in directory `output/` via:
```latex
\documentclass[a4paper]{scrartcl} %twocolumn %scrartcl, scrreprt, scrbook
\usepackage[margin=1in]{geometry}
\usepackage{hyperref}
\usepackage{booktabs}
\usepackage[table]{xcolor}
\usepackage{siunitx}

\setlength{\parindent}{0in}

\newcommand{\RedA}{red}
\newcommand{\RedB}{red!60}
\newcommand{\RedC}{red!30}
\newcommand{\BlueA}{blue!75}
\newcommand{\BlueB}{blue!55}
\newcommand{\BlueC}{blue!30}

\begin{document}
\input{output.tex}
\end{document}
```

## References
thrreg is based on the 
[source code](https://www.ssc.wisc.edu/~bhansen/progs/ecnmt_00.html) 
provided by Bruce E. Hansen.

Hansen, B. E. (2000). Sample splitting and threshold estimation.
*Econometrica*, 68(3):575--603. 
https://onlinelibrary.wiley.com/doi/10.1111/1468-0262.00124. 
Open access: https://www.ssc.wisc.edu/~bhansen/papers/ecnmt_00.pdf.


NEWS.md
# thrreg 0.1.1 (July 28, 2020)

### `thr_est()`
* New argument: `inf.crit`
* Compute additional statistics (adjusted R^2, AIC, BIC, HQC) for linear model
* Compute additional statistics (joint adjusted R^2, joint AIC, joint BIC, 
  joint HQC) for full threshold model



# thrreg 0.1.0 (January 29, 2020)

Updates and changes in functions compared to 
[original source code](https://www.ssc.wisc.edu/~bhansen/progs/ecnmt_00.html) 
by Bruce E. Hansen.

### `thr_test_hom()` and `thr_test_het()`
* New arguments: `var.names`, `cr`, `graph`, `quick`
* A `data.frame` can be passed
* Arguments `yi`, `xi`, `qi` may be of type `integer` or `character`
* Output in LaTeX format 

### `thr_est()`
* New arguments: `test.pvalue`, `conf2`, `nonpar`, `graph`, `signif.level`, 
  `digits`, `integer.digits`, `header`, `output.short`, `signif.legend`
* A `data.frame` can be passed
* Arguments `yi`, `xi`, `qi` may be of type `integer` or `character`
* Loop over different significance levels `c(.9, .95, .99)` for estimation of 
  coefficients
* Definition of notation for significance levels `c(.9, .95, .99)` for 
  threshold test and estimated coefficients
  * for `"stars"`: number of stars
  * for `"colors"`: red/blue tones for positive/negative estimates
* Constructs a summarizing threshold regression table with estimates, 
  standard errors, and significance levels for both regimes
* Changed `"Constant"` to `"Const"` for shorter output in tables
* Output in LaTeX format


test_script.R
library(thrreg)

## Load data
data <- dur_john

## Define variable names for output (optional)
var_names <- c("GDP\\_Gwth", "Log\\_GDP\\_1960", "Log\\_Inv/GDP",
               "Log\\_Pop\\_Gwth", "Log\\_School", "GDP\\_1960", "Literacy")
xi <- 2:5
h <- 1


## First Level
sink("output/output.tex")
cat("\\section{First Sample Split}", "\n\n")

## Test for a Sample Split
# test_hom <- thr_test_hom(data, "GDPGwth", xi, "GDP1960", var_names)
test_1_gdp <- thr_test_het(data, "GDPGwth", xi, "GDP1960", var_names)
test_1_lit <- thr_test_het(data, "GDPGwth", xi, "Literacy", var_names)

## Estimate Sample Split
qhat_1 <- thr_est(data, "GDPGwth", xi, "GDP1960", h, test_1_gdp$p_value,
                  var_names, digits.thr = 0, header = var_names[1])
sink()


## Second Level
dat_2l <- subset(data, GDP1960 <= qhat_1)
dat_2u <- subset(data, GDP1960 > qhat_1)

sink("output/output.tex", append = T)
cat(paste0("\\section{Second Sample Split: Subsample, Incomes below ", qhat_1,
           "}", "\n\n"))

## Test for a Sample Split
test_2l_gdp <- thr_test_het(dat_2l, "GDPGwth", xi, "GDP1960", var_names)
test_2l_lit <- thr_test_het(dat_2l, "GDPGwth", xi, "Literacy", var_names)

cat(paste0("\\section{Second Sample Split: Subsample, Incomes above ", qhat_1,
           "}", "\n\n"))

## Test for a Sample Split
test_2u_gdp <- thr_test_het(dat_2u, "GDPGwth", xi, "GDP1960", var_names)
test_2u_lit <- thr_test_het(dat_2u, "GDPGwth", xi, "Literacy", var_names)

## Estimate Sample Split
qhat_2 <- thr_est(dat_2u, "GDPGwth", xi, "Literacy", h, test_2u_lit$p_value,
                  var_names, digits.thr = 0, header = var_names[1])
sink()


## Third Level
dat_3l <- subset(dat_2u, Literacy <= qhat_2)
dat_3u <- subset(dat_2u, Literacy > qhat_2)

sink("output/output.tex", append = T)
cat(paste0("\\section{Third Sample Split: Subsample, Incomes above ", qhat_1,
           ", Literacy below ", qhat_2, "}", "\n\n"))

## Test for a Sample Split
test_3l_gdp <- thr_test_het(dat_3l, "GDPGwth", xi, "GDP1960")
test_3l_lit <- thr_test_het(dat_3l, "GDPGwth", xi, "Literacy")

cat(paste0("\\section{Third Sample Split: Subsample, Incomes above ", qhat_1,
           ", Literacy above ", qhat_2, "}", "\n\n"))

## Test for a Sample Split
test_3u_gdp <- thr_test_het(dat_3u, "GDPGwth", xi, "GDP1960")
test_3u_lit <- thr_test_het(dat_3u, "GDPGwth", xi, "Literacy")

sink()


R/data.R
#' Durlauf and Johnson (1995) data
#'
#' Data distributed by Durlauf and Johnson (1995) to document their empirical work.
#'
#' @format A data frame with 96 rows and 7 variables:
#' \describe{
#'   \item{GDPGwth}{GDP growth as log difference between per capita GDP in 1985 and 1960.}
#'   \item{LogGDP1960}{Log per capita GDP in 1960.}
#'   \item{LogInvGDP}{Log average ratio of investment (including government investment)
#'     to GDP from 1960 to 1985.}
#'   \item{LogPopGwth}{Log average growth rate of working-age population from 1960 to 1985.}
#'   \item{LogSchool}{Log average fraction of working-age population enrolled in
#'     secondary school from 1960 to 1985.}
#'   \item{GDP1960}{Per capita GDP in 1960.}
#'   \item{Literacy}{Fraction of the population over 15 years old that is able
#'     to read and write in 1960.}
#' }
#'
#' @source
#'     Original data are reported in the appendix of Durlauf and Johnson (1995),
#'     pp. 379--380, and can be downloaded under
#'     \url{http://qed.econ.queensu.ca/jae/1995-v10.4/durlauf-johnson/}.
#'     Original data have been transformed to match Eq. (7) in Durlauf and
#'     Johnson (1995) and Section 5 in Hansen (2000), respectively.
#'     Documentation is taken from the Journal of Applied Econometrics FTP
#'     (see link above).
#'     All of the data with the exception of Literacy are from Mankiw, Romer and
#'     Weil (1992). Literacy is from the World Bank's World Development Report.
#'
#' @references
#'     Durlauf, S. N. and Johnson, P. A. (1995).
#'     Multiple regimes and cross-country growth behavior.
#'     \emph{Journal of Applied Econometrics}, 10(4):365--384.
#'     \url{https://doi.org/10.1002/jae.3950100404}.
#' @references
#'    Mankiw, N. G., Romer, D., and Weil, D. N. (1992).
#'    A contribution to the empirics of economic growth.
#'    \emph{The Quarterly Journal of Economics}, 107(2):407--437.
#'    \url{https://doi.org/10.2307/2118477}.
#'
"dur_john"

#' Original Durlauf and Johnson (1995) data
#'
#' Data distributed by Durlauf and Johnson (1995) to document their empirical work.
#'
#' @format A data frame with 121 rows and 11 variables:
#' \describe{
#'   \item{NUMBER}{Country number in Summers and Heston (1988) dataset.}
#'   \item{NONOIL}{dummy; \code{= 1} for nonoil producing countries.}
#'   \item{INTER}{dummy; \code{= 1} for countries with better quality data.}
#'   \item{OECD}{dummy; \code{= 1} for OECD countries.}
#'   \item{GDP60}{Per capita GDP in 1960.}
#'   \item{GDP85}{Per capita GDP in 1985.}
#'   \item{GDPGRO}{Average growth rate of per capita GDP from 1960 to 1985.}
#'   \item{POPGRO}{Average growth rate of working-age population from 1960 to 1985.}
#'   \item{IONY}{Average ratio of investment (including government investment)
#'     to GDP from 1960 to 1985.}
#'   \item{SCHOOL}{Average fraction of working-age population enrolled in
#'     secondary school from 1960 to 1985.}
#'   \item{LIT60}{Fraction of the population over 15 years old that is able
#'     to read and write in 1960.}
#' }
#'
#' @source
#'     \url{http://qed.econ.queensu.ca/jae/1995-v10.4/durlauf-johnson/}.
#'
#'     Original data are reported in the appendix of Durlauf and Johnson (1995),
#'     pp. 379--380. Documentation is taken from the Journal of Applied
#'     Econometrics FTP (see link above).
#'     All of the data with the exception of LIT60 are from Mankiw, Romer
#'     and Weil (1992). LIT60 is from the World Bank's World Development Report.
#'
#' @note A value of -999 indicates that the observation is missing.


R/thr_test_het.R
#' Threshold Test under Heteroskedasticity
#'
#' Computes a test for a threshold in linear regression under heteroskedasticity.
#'
#' @param quick Integer; indicator of method used for bootstrap.
#'     Set \code{quick = 1} for quick computation of asymptotic distribution.
#'     This method is not a proper bootstrap and may result in excess rejections.
#'     It also uses more memory.
#'     Set \code{quick = 2} (default) for a better bootstrap procedure,
#'     which also uses less memory, but is more time consuming.
#'
#' @inheritParams thr_test_hom
#'
#' @details
#'     \enumerate{
#'     \item Do not include a constant in the independent variables;
#'           the function automatically adds an intercept to the regression.
#'     \item There are two bootstrap methods which the function can use.
#'
#'           The first method, obtained by setting \code{quick = 1},
#'           is the method presented in the paper
#'           Hansen, B. E. (1996). Inference When a Nuisance Parameter is Not
#'           Identified Under the Null Hypothesis.
#'           \emph{Econometrica}, 64(2):413-430.
#'           \url{https://www.ssc.wisc.edu/~bhansen/papers/ecnmt_96.pdf},
#'           which simulates the asymptotic null distribution.
#'           A computational shortcut is also taken which speeds computational
#'           time, at the cost of greater memory usage, so may not be possible
#'           for large data sets.
#'
#'           The second method, obtained by setting \code{quick = 2},
#'           is a "fixed regressor bootstrap", which is quite close.
#'           The difference is that the bootstrap procedure calculates
#'           the variance-covariance matrix in each bootstrap replication.
#'           This results in a better finite sample approximation.
#'           The cost is greater computation time.
#'
#'           The function is set by default to use the second method
#'           (\code{quick = 2}), which has better sampling properties.
#'           If computational time is a concern, switch to the first method
#'           (\code{quick = 1}). If an "out of workspace memory" message appears,
#'           switch back to \code{quick = 2}.
#'     }
#'
#' @inherit thr_test_hom return
#' @inherit thrreg author references
#'
# #' @family threshold regression functions
#'
#' @seealso
#'     \code{\link{thr_test_hom}} for threshold test under homoskedasticity,
#'     \code{\link{thr_est}} for threshold estimation.
#'
#' @keywords htest models ts
#'
#' @examples
#' \donttest{
#' ## Performs part of the empirical work reported in Hansen (2000)
#' data <- dur_john
#' output <- thr_test_het(data, 1, 2:5, 6)
#'
#' output$f_test
#' output$p_value
#' }
#'
# #' @rdname thr_test
#'
#' @importFrom graphics legend lines plot title
#' @importFrom stats rnorm
#'
#' @export
#'
thr_test_het <- function(df, yi, xi, qi, var.names = colnames(df), trim_per = .15,
                         rep = 1000, cr = .95, graph = TRUE, quick = 2) {

  dat <- as.matrix(df)
  if (is.character(yi)) yi <- which(colnames(df) == yi)
  if (is.character(xi)) xi <- which(colnames(df) %in% xi)
  if (is.character(qi)) qi <- which(colnames(df) == qi)

  n <- nrow(dat)
  q <- dat[,qi]
  qs <- order(q)
  y <- as.matrix(dat[qs,yi])
  x <- cbind(matrix(c(1),n,1),dat[qs,xi])
  q <- as.matrix(q[qs])
  k <- ncol(x)
  qname <- var.names[qi]
  qs <- unique(q)
  qn <- length(qs)
  qq <- matrix(c(0),qn,1)
  for (r in 1:qn) qq[r] <- colSums(q==qs[r])
  cqq <- cumsum(qq)
  sq <- (cqq>=floor(n*trim_per))*(cqq<=(floor(n*(1-trim_per))))
  qs <- as.matrix(qs[sq>0])
  cqq <- as.matrix(cqq[sq>0])
  qn <- nrow(qs)

  mi <- solve(t(x)%*%x, tol = 1e-1000)
  e <- y-x%*%mi%*%(t(x)%*%y)
  ee <- t(e)%*%e
  xe <- x*(e%*%matrix(c(1),1,k))
  vi <- t(xe)%*%xe
  cxe <- apply(xe,2,cumsum)
  sn <- matrix(c(0),qn,1)

  if (quick == 1) {
    mmistore <- matrix(c(0),k*(k+1)/2,qn)
    cqqb <- 1
    mm <- matrix(c(0),k,k)
    vv <- matrix(c(0),k,k)
    for (r in 1:qn) {
      cqqr <- cqq[r]
      if (cqqb==cqqr) {
        mm <- mm + as.matrix(x[cqqb,])%*%x[cqqb,]
        vv <- vv + as.matrix(xe[cqqb,])%*%xe[cqqb,]
      } else {
        mm <- mm + t(x[(cqqb:cqqr),])%*%x[(cqqb:cqqr),]
        vv <- vv + t(xe[(cqqb:cqqr),])%*%xe[(cqqb:cqqr),]
      }
      sume <- as.matrix(cxe[cqqr,])
      mmi <- solve(vv-mm%*%mi%*%vv-vv%*%mi%*%mm+mm%*%mi%*%vi%*%mi%*%mm)
      sn[r] <- t(sume)%*%mmi%*%sume
      cqqb <- cqqr+1
      ii <- 1
      for (i in 1:k) {
        mmistore[ii:(ii+i-1),r] <- mmi[i,1:i]
        ii <- ii+i
      }
    }
    si <- which.max(sn)
    qmax <- qs[si]
    lr <- sn
    ftest <- sn[si]
    fboot <- matrix(c(0),rep,1)
    for (j in 1:rep) {
      y  <- rnorm(n)*e
      xe <- x*((y-x%*%mi%*%(t(x)%*%y))%*%matrix(c(1),1,k))
      cxe <- apply(xe,2,cumsum)
      sn <- matrix(c(0),qn,1)
      for (r in 1:qn) {
        mmi <- matrix(c(0),k,k)
        ii <- 1
        for (i in 1:k) {
          mmi[i,1:i] <- mmistore[ii:(ii+i-1),r]
          mmi[1:(i-1),i] <- mmi[i,1:(i-1)]
          ii <- ii+i
        }
        sume <- as.matrix(cxe[cqq[r],])
        sn[r] <- t(sume)%*%mmi%*%sume
      }
      fboot[j] <- max(sn)
    }
  }

  if (quick == 2) {
    cqqb <- 1
    mm <- matrix(c(0),k,k)
    vv <- matrix(c(0),k,k)
    for (r in 1:qn) {
      cqqr <- cqq[r]
      if (cqqb==cqqr) {
        mm <- mm + as.matrix(x[cqqb,])%*%x[cqqb,]
        vv <- vv + as.matrix(xe[cqqb,])%*%xe[cqqb,]
      } else {
        mm <- mm + t(x[(cqqb:cqqr),])%*%x[(cqqb:cqqr),]
        vv <- vv + t(xe[(cqqb:cqqr),])%*%xe[(cqqb:cqqr),]
      }
      sume <- as.matrix(cxe[cqqr,])
      mmi <- vv-mm%*%mi%*%vv-vv%*%mi%*%mm+mm%*%mi%*%vi%*%mi%*%mm
      if (qr(mmi)$rank==ncol(mmi)){
        sn[r] <- t(sume)%*%solve(mmi)%*%sume
      }
      cqqb <- cqqr+1
    }
    si <- which.max(sn)
    qmax <- qs[si]
    lr <- sn
    ftest <- sn[si]
    fboot <- matrix(c(0),rep,1)
    for (j in 1:rep) {
      y  <- rnorm(n)*e
      xe <- x*((y-x%*%mi%*%(t(x)%*%y))%*%matrix(c(1),1,k))
      vi <- t(xe)%*%xe
      cxe <- apply(xe,2,cumsum)
      sn <- matrix(c(0),qn,1)
      cqqb <- 1
      mm <- matrix(c(0),k,k)
      vv <- matrix(c(0),k,k)
      for (r in 1:qn) {
        cqqr <- cqq[r]
        if (cqqb==cqqr) {
          mm <- mm + as.matrix(x[cqqb,])%*%x[cqqb,]
          vv <- vv + as.matrix(xe[cqqb,])%*%xe[cqqb,]
        } else {
          mm <- mm + t(x[(cqqb:cqqr),])%*%x[(cqqb:cqqr),]
          vv <- vv + t(xe[(cqqb:cqqr),])%*%xe[(cqqb:cqqr),]
        }
        mmi <- vv-mm%*%mi%*%vv-vv%*%mi%*%mm+mm%*%mi%*%vi%*%mi%*%mm
        sume <- as.matrix(cxe[cqqr,])
        if (qr(mmi)$rank==ncol(mmi)) {
          sn[r] <- t(sume)%*%solve(mmi)%*%sume
        }
        cqqb <- cqqr+1
      }
      fboot[j] <- max(sn)
    }
  }

  fboot <- as.matrix(sort(fboot))
  pv <- mean(fboot >= matrix(c(1),rep,1)%*%ftest)
  crboot <- fboot[round(rep*cr)]

  if (graph==TRUE) {
    xxlim <- range(qs)
    yylim <- range(rbind(lr,crboot))
    clr <- matrix(c(1),qn,1)*crboot
    plot(qs,lr,lty=1,col=1,xlim=xxlim,ylim=yylim,type="l",ann=0)
    lines(qs,clr,lty=2,col=2)
    title(main=rbind("F Test For Threshold",
                     "Reject Linearity if F Sequence Exceeds Critical Value"),
          xlab="gamma",ylab="Fn(gamma)")
    tit <- paste(cr*100,c("% Critical"),sep="")
    legend("bottomright",c("LRn(gamma)",tit),lty=c(1,2),col=c(1,2))
  }

  cat(paste0("\\subsection{Testing for a Sample Split, Using ", qname, "}"), "\n")
  cat("\\subsubsection*{Test of Null of No Threshold Against Alternative of Threshold}",
      "\n")
  cat("Allowing Heteroskedastic Errors (White Corrected)", "\\\\")
  cat("\\\\\n")
  cat("Number of Bootstrap Replications: ", rep, "\\\\\n")
  cat("Trimming Percentage:              ", trim_per, "\\\\")
  cat("\\\\\n")
  cat("Threshold Variable:                ", qname, "\\\\\n")
  cat("Threshold Estimate:               ", qmax, "\\\\\n")
  cat("LM-test for no threshold:         ", ftest, "\\\\\n")
  cat("Bootstrap P-Value:                ", pv, "\\\\\n")
  cat("\n")

  list(f_test = ftest, p_value = pv)
}


R/thr_test_hom.R
#' Threshold Test under Homoskedasticity
#'
#' Computes a test for a threshold in linear regression under homoskedasticity.
#'
#' @param df Data frame.
#' @param yi Integer or character; index or column name of dependent (y)
#'     variable in \code{df}.
#' @param xi Integer or character vector; indexes or column names of
#'     independent (x) variables in \code{df}.
#' @param qi Integer or character; index or column name of threshold (q)
#'     variable in \code{df}.
#' @param var.names Character vector; variable names with
#'     \code{length(var.names) == ncol(df)}
#'     corresponding to columns in \code{df} to be used in threshold regression
#'     table. Default is \code{colnames(df)}.
#' @param trim_per Numeric; percentage of sample to trim from ends.
#'     Default is \code{trim_per = .15}.
#' @param rep Integer; number of bootstrap replications. Default is
#'     \code{rep = 1000}.
#' @param cr Numeric; confidence level used to plot the critical value in the
#'     graph. It is not used elsewhere in the analysis. Default is
#'     \code{cr = .95}.
#' @param graph Logical; graph indicator.
#'     Set \code{TRUE} (default) to view the graph of the likelihood;
#'     set \code{FALSE} otherwise.
#' @param quick Integer; indicator of method used for bootstrap.
#'     Set \code{quick = 1} (default) for efficient, quick computation;
#'     set \code{quick = 0} if memory is limited (perhaps for large data sets).
#'
#' @details
#'     \enumerate{
#'     \item Do not include a constant in the independent variables;
#'           the function automatically adds an intercept to the regression.
#'     \item The function stores the sequential design matrices in order to
#'           speed up the bootstrap computations. It is possible that if your
#'           dataset is very large, this will exceed your computer RAM memory.
#'           If so, function will crash, and the message
#'           \code{Error: allocMatrix: too many elements specified}
#'           will be displayed. If more RAM is not available, switch to
#'           \code{quick = 0}. The switch \code{quick = 0} requires the bootstrap
#'           to re-calculate the design matrices for each bootstrap replication,
#'           which requires less memory, but somewhat more computer time.
#'     }
#'
#' @return A list with components:
#'     \describe{
#'     \item{\code{f_test}}{the value of Maximal (Quandt) F-statistic.}
#'     \item{\code{p_value}}{the bootstrap p-value.}
#'     }
#'
#' @inherit thrreg author references
#'
# #' @family threshold regression functions   # Automatically links to other functions in package
#'
#' @seealso
#'     \code{\link{thr_test_het}} for threshold test under heteroskedasticity,
#'     \code{\link{thr_est}} for threshold estimation.
#'
#' @keywords htest models ts
#'
#' @examples
#' \donttest{
#' ## Performs part of the empirical work reported in Hansen (2000)
#' data <- dur_john
#' output <- thr_test_hom(data, 1, 2:5, 6)
#'
#' output$f_test
#' output$p_value
#' }
#'
# #' @rdname thr_test
#'
#' @importFrom graphics legend lines plot title
#' @importFrom stats rnorm
#'
#' @export
#'
thr_test_hom <- function(df, yi, xi, qi, var.names = colnames(df), trim_per = .15,
                         rep = 1000, cr = .95, graph = TRUE, quick = 1) {

  dat <- as.matrix(df)
  if (is.character(yi)) yi <- which(colnames(df) == yi)
  if (is.character(xi)) xi <- which(colnames(df) %in% xi)
  if (is.character(qi)) qi <- which(colnames(df) == qi)

  n <- nrow(dat)
  q <- dat[,qi]
  qs <- order(q)
  y <- as.matrix(dat[qs,yi])
  x <- cbind(matrix(c(1),n,1),dat[qs,xi])
  q <- as.matrix(q[qs])
  k <- ncol(x)
  qname <- var.names[qi]
  qs <- unique(q)
  qn <- length(qs)
  qq <- matrix(c(0),qn,1)
  for (r in 1:qn) qq[r] <- colSums(q==qs[r])
  cqq <- cumsum(qq)
  sq <- (cqq>=floor(n*trim_per))*(cqq<=(floor(n*(1-trim_per))))
  qs <- as.matrix(qs[sq>0])
  cqq <- as.matrix(cqq[sq>0])
  qn <- nrow(qs)

  mi <- solve(t(x)%*%x)
  e <- y-x%*%mi%*%(t(x)%*%y)
  ee <- t(e)%*%e
  xe <- x*(e%*%matrix(c(1),1,k))
  xe <- apply(xe,2,cumsum)
  sn <- matrix(c(0),qn,1)

  if (quick == 1){
    mmistore <- matrix(c(0),k*(k+1)/2,qn)
    cqqb <- 1
    mm <- matrix(c(0),k,k)
    for (r in 1:qn) {
      cqqr <- cqq[r]
      if (cqqb==cqqr) {
        mm <- mm + as.matrix(x[cqqb,])%*%x[cqqb,]
      } else { mm <- mm + t(x[(cqqb:cqqr),])%*%x[(cqqb:cqqr),]}
      sume <- as.matrix(xe[cqqr,])
      mmi <- solve(mm - mm%*%mi%*%mm)
      sn[r] <- t(sume)%*%mmi%*%sume
      cqqb <- cqqr+1
      ii <- 1
      for (i in 1:k) {
        mmistore[ii:(ii+i-1),r] <- mmi[i,1:i]
        ii <- ii+i
      }
    }
    si <- which.max(sn)
    smax <- sn[si]
    qmax <- qs[si]
    sig <- (ee - smax)/n
    lr <- sn/as.vector(sig)
    ftest <- smax/sig
    fboot <- matrix(c(0),rep,1)
    for (j in 1:rep) {
      y  <- rnorm(n)
      e  <- y-x%*%mi%*%(t(x)%*%y)
      ee <- t(e)%*%e
      xe <- x*(e%*%matrix(c(1),1,k))
      xe <- apply(xe,2,cumsum)
      sn <- matrix(c(0),qn,1)
      for (r in 1:qn) {
        mmi <- matrix(c(0),k,k)
        ii <- 1
        for (i in 1:k) {
          mmi[i,1:i] <- mmistore[ii:(ii+i-1),r]
          mmi[1:(i-1),i] <- mmi[i,1:(i-1)]
          ii <- ii+i
        }
        sume <- as.matrix(xe[cqq[r],])
        sn[r] <- t(sume)%*%mmi%*%sume
      }
      smax <- max(sn)
      sig <- (ee - smax)/n
      fboot[j] <- smax/sig
    }
  }

  if (quick == 0) {
    cqqb <- 1
    mm <- matrix(c(0),k,k)
    for (r in 1:qn) {
      cqqr <- cqq[r]
      if (cqqb==cqqr) {
        mm <- mm + as.matrix(x[cqqb,])%*%x[cqqb,]
      } else { mm <- mm + t(x[(cqqb:cqqr),])%*%x[(cqqb:cqqr),]}
      sume <- as.matrix(xe[cqqr,])
      mmi <- mm - mm%*%mi%*%mm
      sn[r] <- t(sume)%*%solve(mmi)%*%sume
      cqqb <- cqqr+1
    }
    si <- which.max(sn)
    smax <- sn[si]
    qmax <- qs[si]
    sig <- (ee - smax)/n
    lr <- sn/as.vector(sig)
    ftest <- smax/sig
    fboot <- matrix(c(0),rep,1)
    for (j in 1:rep) {
      y  <- rnorm(n)
      e  <- y-x%*%mi%*%(t(x)%*%y)
      ee <- t(e)%*%e
      xe <- x*(e%*%matrix(c(1),1,k))
      xe <- apply(xe,2,cumsum)
      sn <- matrix(c(0),qn,1)
      cqqb <- 1
      mm <- matrix(c(0),k,k)
      for (r in 1:qn) {
        cqqr <- cqq[r]
        if (cqqb==cqqr) {
          mm <- mm + as.matrix(x[cqqb,])%*%x[cqqb,]
        } else { mm <- mm + t(x[(cqqb:cqqr),])%*%x[(cqqb:cqqr),]}
        mmi <- mm - mm%*%mi%*%mm
        sume <- as.matrix(xe[cqq[r],])
        sn[r] <- t(sume)%*%solve(mmi)%*%sume
        cqqb <- cqqr+1
      }
      smax <- max(sn)
      sig <- (ee - smax)/n
      fboot[j] <- smax/sig
    }
  }

  fboot <- as.matrix(sort(fboot))
  pv <- mean(fboot >= matrix(c(1),rep,1)%*%ftest)
  crboot <- fboot[round(rep*cr)]

  if (graph==TRUE) {
    xxlim <- range(qs)
    yylim <- range(rbind(lr,crboot))
    clr <- matrix(c(1),qn,1)*crboot
    plot(qs,lr,lty=1,col=1,xlim=xxlim,ylim=yylim,type="l",ann=0)
    lines(qs,clr,lty=2,col=2)
    title(main=rbind("F Test For Threshold",
                     "Reject Linearity if F Sequence Exceeds Critical Value"),
          xlab="gamma",ylab="Fn(gamma)")
    tit <- paste(cr*100,c("% Critical"),sep="")
    legend("bottomright",c("LRn(gamma)",tit),lty=c(1,2),col=c(1,2))
  }

  cat(paste0("\\subsection{Testing for a Sample Split, Using ", qname, "}"), "\n")
  cat("\\subsubsection*{Test of Null of No Threshold Against Alternative of Threshold}",
      "\n")
  cat("Under Maintained Assumption of Homoskedastic Errors", "\\\\")
  cat("\\\\\n")
  cat("Number of Bootstrap Replications: ", rep, "\\\\\n")
  cat("Trimming Percentage:              ", trim_per, "\\\\")
  cat("\\\\\n")
  cat("Threshold Variable:                ", qname, "\\\\\n")
  cat("Threshold Estimate:               ", qmax, "\\\\\n")
  cat("F-test for no threshold:          ", ftest, "\\\\\n")
  cat("Bootstrap P-Value:                ", pv, "\\\\\n")
  cat("\n")

  list(f_test = as.vector(ftest), p_value = pv)
}


R/thrreg.R
#' thrreg: Threshold Regression Model
#'
#' thrreg estimates the threshold regression model by Hansen (2000).
#' It is based on the
#' \href{https://www.ssc.wisc.edu/~bhansen/progs/ecnmt_00.html}{source code}
#' provided by Bruce E. Hansen.
#'
#' @details thrreg provides three functions:
#'     \itemize{
#'     \item \code{\link{thr_test_hom}} tests for a threshold in a linear
#'           regression under homoskedasticity.
#'     \item \code{\link{thr_test_het}} tests for a threshold in a linear
#'           regression under heteroskedasticity.
#'     \item \code{\link{thr_est}} estimates the threshold and the regression
#'           parameters of the threshold model.
#'     }
#'
#' @author Marcel Kremer, \email{marcel.kremer@@uni-due.de}
#' @author Bruce E. Hansen, \email{behansen@@wisc.edu}
#'
#' @references
#'     Hansen, B. E. (2000). Sample splitting and threshold estimation.
#'     \emph{Econometrica}, 68(3):575--603.
#'     \url{https://doi.org/10.1111/1468-0262.00124}.
#'     \url{https://www.ssc.wisc.edu/~bhansen/papers/ecnmt_00.pdf}.
#'
#' @seealso Useful links:
#'     \itemize{
#'     \item \url{https://github.com/mlkremer/thrreg}
#'     \item Report bugs at \url{https://github.com/mlkremer/thrreg/issues}
#'     }
#'
#' @examples ## See README.md on GitHub for a comprehensive example.
#'
#' @docType package
#' @name thrreg
NULL
